File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/service/SerialExecutor.java
package com.battlelancer.seriesguide.service;

import android.os.AsyncTask;
import android.support.annotation.NonNull;
import java.util.ArrayDeque;
import java.util.concurrent.Executor;

/**
 * A serial executor that uses the {@link AsyncTask#THREAD_POOL_EXECUTOR} to execute tasks on.
 *
 * <p>Adapted from {@link AsyncTask#SERIAL_EXECUTOR}, Copyright (C) 2008 The Android Open Source
 * Project, Licensed under the Apache License, Version 2.0.
 */
class SerialExecutor implements Executor {
    private final ArrayDeque<Runnable> tasks = new ArrayDeque<>();
    private Runnable active;

    public synchronized void execute(@NonNull final Runnable r) {
        tasks.offer(new Runnable() {
            public void run() {
                try {
                    r.run();
                } finally {
                    scheduleNext();
                }
            }
        });
        if (active == null) {
            scheduleNext();
        }
    }

    private synchronized void scheduleNext() {
        if ((active = tasks.poll()) != null) {
            AsyncTask.THREAD_POOL_EXECUTOR.execute(active);
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/service/SerialExecutor.kt
package com.battlelancer.seriesguide.service

import android.os.AsyncTask
import java.util.ArrayDeque
import java.util.concurrent.Executor

/**
 * A serial executor that uses the [AsyncTask.THREAD_POOL_EXECUTOR] to execute tasks on.
 *
 * Adapted from [AsyncTask.SERIAL_EXECUTOR], Copyright (C) 2008 The Android Open Source
 * Project, Licensed under the Apache License, Version 2.0.
 */
class SerialExecutor : Executor {
    private val tasks = ArrayDeque<Runnable>()
    private var active: Runnable? = null

    @Synchronized
    override fun execute(r: Runnable) {
        tasks.offer(Runnable {
            try {
                r.run()
            } finally {
                scheduleNext()
            }
        })
        if (active == null) {
            scheduleNext()
        }
    }

    @Synchronized
    private fun scheduleNext() {
        if (tasks.poll().also { active = it } != null) {
            AsyncTask.THREAD_POOL_EXECUTOR.execute(active)
        }
    }
}